analysis/src/python/data_analysis/template_errors/postprocessing.py

```diff
@@ -0,0 +1,162 @@
+    biggest_threshold = max_template_freq / 100
+    # Filter issues with frequency les than 10%
+    filtered = templates_df.loc[templates_df[SubmissionColumns.FREQUENCY.value] > lowest_threshold]
+    template_issues = filtered[filtered[SubmissionColumns.FREQUENCY.value] > biggest_threshold]
+    rare_typical_issues = filtered[(filtered[SubmissionColumns.FREQUENCY.value] > lowest_threshold) & (
+            filtered[SubmissionColumns.FREQUENCY.value] <= middle_threshold)]
+    common_typical_issues = filtered[(filtered[SubmissionColumns.FREQUENCY.value] > middle_threshold) & (
+            filtered[SubmissionColumns.FREQUENCY.value] <= biggest_threshold)]
+    return template_issues, rare_typical_issues, common_typical_issues
+
+
+def find_user_solutions(
+        template_issues_df: pd.DataFrame,
+        raw_issues_df: pd.DataFrame,
+        output_path: str,
+        n: int = 5,
+        to_add_description: bool = True,
+) -> pd.DataFrame:
+    base_path = f'{output_path}/solutions'
+    description = 'description'
+    final_descriptions = []
+    for _, row in template_issues_df.iterrows():
+        step_rows = raw_issues_df.loc[
+            raw_issues_df[SubmissionColumns.STEP_ID.value] == row[SubmissionColumns.STEP_ID.value]]
+        current_attempt = 0
+        visited_rows = []
+        wrong_rows = []
+        descriptions = []
+        cur_path = f'{base_path}/{row[SubmissionColumns.STEP_ID.value]}/{row[SubmissionColumns.RAW_ISSUE_CLASS.value]}'
+        create_directory(cur_path)
+        while current_attempt < n:
+            row_number = random.randint(0, step_rows.shape[0] - 1)
+            if row_number in visited_rows or row_number in wrong_rows:
+                continue
+            issues = json.loads(step_rows[SubmissionColumns.RAW_ISSUES.value].iloc[row_number], cls=RawIssueDecoder)
+            current_issues = list(
+                filter(lambda i: i.origin_class == row[SubmissionColumns.RAW_ISSUE_CLASS.value], issues))
+            if len(current_issues) == 0:
+                wrong_rows.append(row_number)
+                continue
+            file_path = Path(f"{cur_path}/solution_{current_attempt}.java")
+            next(create_file(file_path, step_rows["code"].iloc[row_number]))
+            visited_rows.append(row_number)
+            descriptions.append(current_issues[random.randint(0, len(current_issues) - 1)].description)
+            current_attempt += 1
+        if to_add_description:
+            final_descriptions.append(descriptions[random.randint(0, len(descriptions) - 1)])
+    template_issues_df[description] = final_descriptions
+    return template_issues_df
+
+
+def add_additional_info(
+        template_issues_df: pd.DataFrame,
+        output_path: str,
+        raw_issues_df: Optional[pd.DataFrame] = None,
+) -> pd.DataFrame:
+    template_issues_df['task_link'] = template_issues_df.apply(
+        lambda row: f'https://hyperskill.org/learn/step/{row[SubmissionColumns.STEP_ID.value]}', axis=1)
```

Comment:
Maybe somehow get from dataset with steps of unify with [`analysis/src/python/data_collection/hyperskill/api/steps.py`](https://github.com/hyperskill/hyperstyle-analyze/blob/21db5d9d14e8e83d946f2d81c8249c6561d005ea/analysis/src/python/data_collection/hyperskill/api/steps.py#L103)